README.md
# Paradisa

[![Version](https://img.shields.io/badge/Version-1.0.0-success.svg)](http://focs.ji.sjtu.edu.cn:2143/projects/team-18/repository)
[![Status](https://img.shields.io/badge/Status-Release-success.svg)](http://focs.ji.sjtu.edu.cn:2143/projects/team-18/repository)

Please visit [here](https://lighghteeloo.github.io/ParadisaRelease/) with a desktop device to play : )

*Paradisa* is the heroine's a girls' name. 

> According to *thinkbabynames.com*,
>
> > *Paradisa* is of Greek origin, and the meaning of Paradisa is "garden, orchard". Paradise is the ancient term for the home of the blessed. It refers to the Garden of Eden in the Old Testament and heaven in the New Testament.
>
> The character imagines her house as a wonderland, or, paradise. The title is also implying her paranoia.

## Table of Contents
- [Background](#Story)
- [Install](#Install)
- [Usage](#Usage)
- [Contributors](#Contributors)

## Story

### Idea

- The house is a big palace.

- My house is a big palace, with caves, zoos, forests and Antic Ocean, where I eat, I practice, I wash. 

- The house is a big palace, with a dark shadow in the deep, howling from time to time.

- My house is a big palace, without a proper exit.

- ...

The story is about the expedition of a paranoid girl called Paradisa in her own house, with her loving kitty.
Though her aim is to do the most daily and normal things, she imagined them as fantasy. Her mother has gone,
and her father is who she afraid of. The only one she can rely on is her cat. So what would happen in this normal day? 

### Content

Paradisa is a typical puzzle-solving side scroller with multiple levels.
For each level, the little girl is aiming at reaching a particular position on the scene and collect
the particular "key item". There are hidden items in some levels, which means the item that is not in
the right path to reach the goal, but may be discovered by careful players. The cat is the companion
and the guidance of the little girl, who resembles the doll that mom left, and a metaphor for mom.
After four levels of collecting, the little girl would come to the final stage, in which she is chased
by a dark shadow, her father. She needs to run higher and higher onto the tower until she falls.

### Mechanism

Although Paradisa is a one-player game, there are two characters in the scene, the little girl and a cat, and 
player can control either the cat or the girl with `← ↑ → ↓` or `A W D S`, and use `C` to change the character.
The little girl can only walk towards `left` or `right`, she will walk when an instruction is given,
and she won't stop until she reaches the end or boundary. Though the cat can freely walk around the scene,
it is not high enough or heavy enough for some objects, so it has to be cradled by the little girl to pass through.
Two characters must get to the target place together and collect the key item to pass the level.
Some hidden items along the way would influence the ending players reach.

### Levels

#### 1. Menu

2D map on the house, available to choose the level

#### 2. Living room (zoo) - green and yellow

*For tutorial* 

key: TV remote control

#### 3. Bedroom (cloud and amusement park) - pink and white

key: pillow

hidden: cat doll given by mom

#### 4. Bathroom (ocean/aquarium) - blue and black

key: toothbrush

#### 5. Storeroom (cave) - black and yellow

key: violin

hidden: music box as a gift from father

#### 6. Balcony (tower) - gray

Father (black shadowed slime chasing beneath)

Choose routine and quickly climb up the tower

To the highest place: 
- The girl would jump from the tower if no hidden item found
- Otherwise, father would become human and they would hug each other.

## Install

Will be updated in later version.

## Usage

- Global:
    - **Pause**: `SAPCE`
    - **Interact**: `E`
    - **God** mode: `G`
    - **Skip** the level: `P`
    - **Restart** the level: `R`
- Cat:
    - Jump: `↑` or `W`
    - Move Left: `←` or `A`
    - Move Right: `→` or `D`
    - Change control to *Girl*: **HOLD** `C`
- Girl:
    - Ladder Up: `↑` or `W`
    - Ladder Up: `↓` or `S`
    - Move Left: `←` or `A`
    - Move Right: `→` or `D`
    - Change control to *Cat*: **RELEASE** `C`

## Contributors
- Rundong Tang
- Yuchen Jiang
- Yuchen Zhou
- Zhimin Sun


ChangeLog.md
# Change Log

## Global

2020-07-13 19:47 	Yuchen Jiang

[MVP] #167 Keep readying...

- Minor Change to Changelog (including this one)
- Petting cat.

2020-07-13 19:00 	Zhimin Sun	

[MVP] #167 Ready to release
- Add README
- Add ChangeLog

2020-07-13 19:08	Yuchen Jiang	

[Fix Physics] #167 

- Block is working. Fixed the bug of pushing away the object too far away.
- Removed the softBlock function. It was a stupid idea.
- Right Surround still wrong...

2020-07-13 18:26	Yuchen Jiang	

[Fix] #167 

- Cat Block size fix
- offset in Image
- Pause done

2020-07-13 11:53 	Yuchen Jiang 	

[src TidyUp] #167 
- Tools.elm -> StringTools.elm
- Character.elm removed
- Look.elm tidied

2020-07-13 06:58 	Yuchen Jiang 	

[Makefile] #167 
- make to build
- css working

2020-07-13 01:21 	Yuchen Jiang 	

[Message] #167 
- Message changed to fit in.
- init fixed.
- update fixed.

2020-07-13 00:36 	Yuchen Jiang 	

[Fix] #167 
- Changed Comment Bug
- Update minor fix

2020-07-13 00:20 	Yuchen Jiang 	

[Update Stringed] #167 
- UpdateLogic
- UpdatePhysics
- UpdateVisual

## Elapse

2020-07-13 09:00 	Yuchen Jiang
	
[Elapse v1 1/1] #167 
- elapse based physics
- elapse based visualization

2020-07-13 07:15 	Yuchen Jiang 		

[Elapse v0 1/1] #167 
- added elapse to slow down animation.
- dummy -> default in Model.elm

## Resize

2020-07-12 15:01 	Yuchen Jiang 	

[Resize v0 2/2] #375 Main changed.
- Resize failing from js side.
- Used % in all view.

2020-07-12 01:38 	Yuchen Jiang 	

[Resize v0 1/2] #375
- Try Stopping resizing. Test needed.
- .gitignore updated.

## Images

2020-07-12 23:28 	Yuchen Zhou 	

[Images v0 2/2] #248 
- Reality in Level1

2020-07-12 01:17 	Yuchen Zhou

[Images v0 1/2] #248
- images of level1

## Operation

2020-07-11 20:49 	Yuchen Jiang 

[Operation] #167 
- type Control in GameModel changed.

## Logic

2020-07-12 00:56 	Zhimin Sun 	

[Logic v1 1/1] #260 Update functions & related

- New functions: Use Msg to generate status.
- Delete the Operation type, edit the status in "CatGirl.elm".

2020-07-10 16:53 	Zhimin Sun 	

[Logic v0 1/1] #260 Update functions & related
- Complete update function: Use Msg to generate Operation.
- Rewrite the Operation type.
- Add new key for `restart` and `skip`.

## Key


2020-07-10 09:36 	Yuchen Jiang 	
[Key v0 2/2] #167 Add Operation.

2020-07-10 00:25 	Yuchen Jiang 	
[Key v0 1/2] #167 Add key.

## View

2020-07-13 05:01 	Yuchen Jiang 	

[View] #167 
- new View function
- Flipping cat
- CatWalk operational.
- Todo: CatJump
- index.js fix

2020-07-13 01:28 	Rundong Tang 	

visualize level 1 images

2020-07-13 01:04 	Yuchen Jiang 	

[Artwork] #167 
- Move images.

2020-07-13 00:45 	Yuchen Zhou 
	
[Visual] #191 ViewInit changes.

2020-07-13 00:42 	Rundong Tang 	

[View] #167 
- add the vlzImg function to View
- View localized

2020-07-13 00:39 	Rundong Tang 	

[View] #167 
- add the vlzImg function to View

2020-07-13 00:28 	Yuchen Jiang 	

[Visualization Fix] #191 
- Changed CatStatus
- Fixed type alteration

2020-07-12 01:17 	Yuchen Zhou 	

[Visualization] #191 

- Link it with URLs.
- Add view-init for Model.elm
- Change the name of some pictures.
- Fix ladder for "List Image".

2020-07-12 00:16 	Rundong Tang 	

[View] #261 add the vlzImg function to View

2020-07-11 11:19 	Yuchen Zhou 	

[Visual] #191 Visualization Sketch done!
- VisualCat and VisualGirl done.
- Visualization done.
- VisualStatic separated.

2020-07-10 14:14 	Rundong Tang 
	
[View] #261 frame of View

2020-07-10 04:52 	Yuchen Zhou

[Visual] #191 
- imageLoop
- transmitSize
- Orientation

2020-07-09 20:48 	Yuchen Zhou

[View Model] # 191
- Add some link from status to images
- Add many images
- Need sync with gameModel

## Physics

2020-07-12 20:19 	Yuchen Jiang 	

[Physics v0 9/9] #189 
- updatePhysics completed!
- just an MVP implement.
- Used block thinking. Next stage is lining them.

2020-07-12 20:19 	Yuchen Jiang 	

[Physics v0 8/9] #189 Accelerate
- Accelerate complete
- ClearAccelerate complete.
- Ready to upadtePhysics.

2020-07-11 21:18 	Yuchen Jiang 	

[Physics v0 7/9] #189 
- updatePhysics halfway

2020-07-11 15:50 	Yuchen Jiang 	

[Physics v0 6/9] #189 Accelerate
- Accelerate complete
- ClearAccelerate complete.
- Ready to upadtePhysics.

2020-07-11 12:34 	Yuchen Jiang 	

[Physics v0 5/9] #189 Collide
- Use Collide to do collision
- Next is to Accelerate

2020-07-11 09:00 	Yuchen Jiang 	

[Physics v0 4/9] #189 Tidy up Types.
- Sort block-related functions into PhysicsTypes.elm

2020-07-11 07:56 	Yuchen Jiang 	

[Physics v0 3/9] #189 Blocks.
- the idea of soft blocks.
- surrounding blocks.
- basic functions about blocks.

2020-07-10 00:25 	Yuchen Jiang 	

[Physics v0 2/9] #189 GameModel change. Master merge wait.

2020-07-09 22:54 	Yuchen Jiang

[Physics v0 1/9] #189 GameModel change. Master merge wait.

## Annotator

2020-07-07 19:44 	Yuchen Zhou 	

[Annotator v0 1/1] See Test.elm.

## Character

2020-07-07 22:21 	Zhimin Sun

[Character v0 2/2] Add character view

2020-07-07 19:37 	Yuchen Zhou

[Character v0 1/2] Walking Cat

## Framework

2020-07-10 02:28 	Yuchen Zhou 	

[Framework v1 12/12] #167 Physics prepare
- Fixed physics.
- VisualType split.
- Change the Resize and GetViewPort :-)

2020-07-09 22:05 	Yuchen Jiang

[Framework v1 11/12] #167 

- BasicTypes.elm: structural change. Physics prepare.

2020-07-09 22:05 	Yuchen Jiang

[Framework v1 10/12] #167 New Test.

2020-07-09 17:19 	Yuchen Jiang 	

[Framework v1 9/12] BasicTypes.elm: structural change. Physics prepare.

2020-07-09 17:19 	Yuchen Jiang 	

[Framework v1 8/12] New Test.

2020-07-08 23:59 	Yuchen Jiang 	

[Framework v1 7/12] BasicTypes.elm: structural change. Physics prepare.


preload.js
const preload = function() {
    setTimeout(function() {
        // Cat
        new Image().src = "./cat/jumping 1.png";
        new Image().src = "./cat/running 1.png";
        new Image().src = "./cat/running 2.png";
        new Image().src = "./cat/running 3.png";
        new Image().src = "./cat/running 4.png";
        new Image().src = "./cat/running 5.png";
        new Image().src = "./cat/running 6.png";
        new Image().src = "./cat/running 7.png";
        new Image().src = "./cat/running 8.png";
        new Image().src = "./cat/sitting 1.png";
        new Image().src = "./cat/sitting 2.png";
        new Image().src = "./cat/sitting 3.png";
        new Image().src = "./cat/sitting 4.png";
        new Image().src = "./cat/sitting 5.png";
        new Image().src = "./cat/sitting 6.png";
        new Image().src = "./cat/sitting 7.png";
        new Image().src = "./cat/standing 1.png";
        // Girl
        new Image().src = "./girl/climbing 1.png";
        new Image().src = "./girl/climbing 2.png";
        new Image().src = "./girl/climbing 3.png";
        new Image().src = "./girl/floating 1.png";
        new Image().src = "./girl/floating 2.png";
        new Image().src = "./girl/floating 3.png";
        new Image().src = "./girl/floating 4.png";
        new Image().src = "./girl/floating 5.png";
        new Image().src = "./girl/floating 6.png";
        new Image().src = "./girl/floating 7.png";
        new Image().src = "./girl/floating 8.png";
        new Image().src = "./girl/floating 9.png";
        new Image().src = "./girl/floating 10.png";
        new Image().src = "./girl/floating 11.png";
        new Image().src = "./girl/floating 12.png";
        new Image().src = "./girl/hugging 1.png";
        new Image().src = "./girl/staying 1.png";
        new Image().src = "./girl/walking 1.png";
        new Image().src = "./girl/walking 2.png";
        new Image().src = "./girl/walking 3.png";
        new Image().src = "./girl/walking 4.png";
        new Image().src = "./girl/walking 5.png";
        // L0
        new Image().src = "./L0/menu.png";
        new Image().src = "./L0/menu1.png";
        new Image().src = "./L0/menu2.png";
        new Image().src = "./L0/menu3.png";
        new Image().src = "./L0/menu4.png";
        // L1
        new Image().src = "./L1_scene/box.png";
        new Image().src = "./L1_scene/grass.png";
        new Image().src = "./L1_scene/item.png";
        new Image().src = "./L1_scene/ladder3.png";
        new Image().src = "./L1_scene/reality.png";
        new Image().src = "./L1_scene/scene.png";
        // L2
        new Image().src = "./L2_scene/box.png";
        new Image().src = "./L2_scene/item.png";
        new Image().src = "./L2_scene/reality.png";
        new Image().src = "./L2_scene/scene.png";
        // L3
        new Image().src = "./L3_scene/box.png";
        new Image().src = "./L3_scene/box2.png";
        new Image().src = "./L3_scene/bubble.png";
        new Image().src = "./L2_scene/reality.png";
        new Image().src = "./L3_scene/lift.png";
        new Image().src = "./L3_scene/control.png";
        new Image().src = "./L3_scene/scene.png";
        // L4
        new Image().src = "./L4_scene/crystal1.png";
        new Image().src = "./L4_scene/crystal2.png";
        new Image().src = "./L4_scene/lift.png";
        new Image().src = "./L2_scene/reality.png";
        new Image().src = "./L4_scene/lowland.png";
        new Image().src = "./L4_scene/scene.png";
        new Image().src = "./L4_scene/light.png";
        new Image().src = "./L4_scene/violin.png";
        new Image().src = "./L4_scene/bow.png";
        new Image().src = "./L4_scene/bow_candle.png";
        // L5
        new Image().src = "./L5_scene/background.png";
        new Image().src = "./L5_scene/ladders.png";
        new Image().src = "./L5_scene/light.png";
        new Image().src = "./L5_scene/reality.png";
        new Image().src = "./L5_scene/reality1.png";
        new Image().src = "./L5_scene/reality2.png";
        new Image().src = "./L5_scene/tower.png";
        new Image().src = "./L5_scene/cat.png";
        // Monster
        new Image().src = "./monster/1.png";
        new Image().src = "./monster/2.png";
        new Image().src = "./monster/3.png";
        new Image().src = "./monster/4.png";
        new Image().src = "./monster/5.png";
        new Image().src = "./monster/6.png";
        new Image().src = "./monster/7.png";
        new Image().src = "./monster/8.png";
        new Image().src = "./monster/9.png";
        new Image().src = "./monster/10.png";
        new Image().src = "./monster/11.png";
        new Image().src = "./monster/12.png";
        new Image().src = "./monster/13.png";
        new Image().src = "./monster/14.png";
        new Image().src = "./monster/15.png";
        new Image().src = "./monster/16.png";
        new Image().src = "./monster/17.png";
        new Image().src = "./monster/18.png";
        new Image().src = "./monster/19.png";
        new Image().src = "./monster/20.png";
        // music TODO: Add music
        new Audio().src = "./music/L0.mp3";
        new Audio().src = "./music/L1.mp3";
        new Audio().src = "./music/L2.mp3";
        new Audio().src = "./music/L3.mp3";
        new Audio().src = "./music/L4.mp3";
        new Audio().src = "./music/L5.mp3";
        new Audio().src = "./music/L6.mp3";
    }, 1000);
};
